fecgov/openFEC
# fix cand election yr ofec candidate history mv

## Summary (required)

- Resolves #3736
Fix cand_election_yr in ofec_candidate_history_mv

the calculation here for the election_year is similar to the election_yr_to_be_included column in ofec_cand_cmte_linkage table, but not exactly the same.  
The purpose of election_yr_to_be_included is to calculated the election_cycle the cycle financial data belongs to and an odd year will be rounded up and folded into its even year cycle.  
Here the election_year is for the election year, an odd election year need to be preserved, separated from its even year cycle.

The calculation is based on the fec_election_yr, the cand_election_yr, the cycle length (6, 4, 2), the next_election 
Simple and straight foward rules first, followed by more complicated rules.  
cycles, election_years, election_districts, active_through
are still based on the original cand_election_yr from disclosure.cand_valid_fec_yr so the interface will not be impacted.  Only candidate_election_year information is updated 

## How to test the changes locally
-Download the branch feature/3736-Fix-cand_election_yr-ofec_candidate_history_mv, run migration file V0135 (or run invoke create_sample_db) to make sure migration files runs without error.

tmp materialized views and their counter part vw_tmp_jj had been created in DEV database that has the updated MVs logic.
(Note: ofec_candidate_totals_mv_tmp_jj does not change in this ticket, but it references ofec_cand_cmte_linkage_vw_tmp_jj and ofec_candidate_history_with_future_election_vw_tmp_jj)
public.ofec_candidate_history_mv_tmp_jj
public.ofec_candidate_history_with_future_election_mv_tmp_jj
public.ofec_cand_cmte_linkage_mv_tmp_jj
public.ofec_candidate_totals_mv_tmp_jj

To test database:
(Cand_id:
-- 'P60007168': Bernie Sandersn (runs 2016, 2020)
make sure 2018 cycle financial data is calculated as part of election_cycle 2020, not 2016)
-- 'P00003392': Hillary Clinton (runs 2008, 2016)
make sure 2014 cycle financial data is calculated as part of election_cycle 2016, not 2008; 2010, 2012, 2018, and 2020 cycle data should not be calculated as part of neither election_cycle 2008 nor 2016.
-- 'H0AL03101': has special election in odd year, has gaps in between two election_cycles

--Connect to DEV database, testing any of the following examples.

select candidate_id, 
two_year_period,
candidate_election_year, 
election_years,
election_districts,
active_through
from ofec_candidate_history_mv_tmp_jj
--where candidate_id = 'H0IL10120'
--where candidate_id = 'P80005960'
-- Clinton
--where candidate_id = 'P00003392'
--where candidate_id = 'S0NY00188'
-- Sanders
--where candidate_id = 'P60007168'
-- Special Election
--where candidate_id = 'S8MN00578'
where candidate_id = 'H0AL03101'
--where candidate_id = 'S4TX00086'
--where candidate_id = 'H4TX19136'
--where candidate_id = 'S8VA00016'
order by 1, 2, 3

- To test API:
On your local machine

In models/candidates.py file, make the following changes:
In class CandidateHistory(BaseCandidate):
    __tablename__ = 'ofec_candidate_history_mv_tmp_jj'

In class CandidateHistoryWithFuture(BaseCandidate):
    __tablename__ = 'ofec_candidate_history_with_future_election_mv_tmp_jj'

In class CandidateTotal(db.Model):
    __tablename__ = 'ofec_candidate_totals_mv_tmp_jj'

In class Candidate:
replace ofec_cand_cmte_linkage_mv with ofec_cand_cmte_linkage_mv_tmp_jj in
secondaryjoin=
secondaryjoin=
order_by=

In models/links/py file, make the following changes:
In class CandidateCommitteeLink:
tablename = 'ofec_cand_cmte_linkage_mv_tmp_jj'

Point to the DEV database and start the local server.
For /candidate/{candidate_id}/history/ endpoint:
candidate_id: P60007168
candidate_id: P00003392
candidate_id:  H0AL03101

For /candidate/totals endpoint:
candidate_id: P60007168
election_year: 2020
election_full: true

candidate_id: P00003392
election_year: 2016
election_full: true

For /elections/ endpoint:
cycle: 2020
election_full: true
office: president
(bernie sanders is the second entry, with total receipts = 20817691.1)

cycle: 2016
election_full: true
office: president
(Hillary Clinton is the first entry, with total receipts = 585699061.27)



## Impacted areas of the application
List general components of the application that this PR will affect:

There is no API change in this ticket.  Output data may change due to the underline MV changes.
-  /candidate/{candidate_id}/history/
- /candidate/totals
- /candidate/totals

## Related PRs
List related PRs against other branches:

branch | PR
------ | ------
fix/other_pr | [link]()
feature/other_pr | [link]()
